atc/engine/builder/delegate_factory.go

```diff
@@ -357,9 +367,15 @@ func (writer *dbEventWriter) Write(data []byte) (int, error) {
 
 	writer.dangling = nil
 
+	payload := string(text)
+	if writer.filter != nil {
+		fmt.Println("EVAN: going through filter: %s", payload)
+		payload = writer.filter(payload)
```

Comment:
Filtering here may not always work as any credential could end up being split into separate `Write()` calls. This is more likely for multi-line credentials like SSH keys, but theoretically it could happen at any time, since `Write()` receives arbitrary chunks of output based largely on timing.

This aspect feels like a tricky problem to solve. :thinking:

A few ideas:

* Watch for segments of credentials instead of the entire value. For example, maybe for multi-line values we could redact each line independently, no matter where it shows up? This could also aid in cases where a multiline value is printed in a marshalled form, e.g. `"foo\nbar"` would result in `"[redacted **x**]\n[redacted **x**]"`.
* Watch out for things that match a prefix of a credential (I suppose up to some threshold). This would likely require buffering somewhere. Maybe there could be an `io.Writer` that "holds on" to the buffer until it doesn't match the prefix of any configured credentials?
* Both?

Reply:
I have an idea. For those multi-line secrets, can we just redact few characters for each line? Which would significantly mitigate random chunk problem.

Reply:
Not sure what you mean - do you mean 'chunk' the values that we redact, too? so e.g. 'abcdefghijklmnop' would redact 'abc', 'def', 'ghi', 'jkl' etc. individually?

Reply:
Sorry for the confusion. Let me be clearer. Say a multi-line ssh key like:

     dkjffkfgkgusjdjjdjshhsjsjsjjshdhddjdhdhdhdjd
     skfjfjekjfk834899rfddf89823893282382303
     slsl00090809009fdjsdkjelkfdjkjdllsdlkklsdkld
    ...

We don't have redact the entire ssh key, as long as we redact partial of it, such as redacting first 8 chars of each line, then the key will lose its meaning:

     [**redacted**]sjdjjdjshhsjsjsjjshdhddjdhdhdhdjd
     [**redacted**]34899rfddf89823893282382303
     [**redacted**]ere009fdjsdkjelkfdjkjdllsdlkklsdkld
    ...

This way, as long as there is 1 line is redacted, the key lose its meaning, so that it significantly reduced possibility of not redacting due to random chunks of build logs.

From implementation perspective, we can randomly choose some pieces from a long secrets to redact.

Reply:
Ah, gotcha. I don't think users would be comfortable with any amount of information leakage as it weakens the entropy of the secret. And if we randomly remove pieces it would potentially reveal the full secret over enough runs.

Reply:
Ok, let's leave this discussion to a separate PR.